README
                        _              _       _          _
   _ __    __ _  ___   | |_ ___   ___ | |___  | | ___  __| |
  | '__|  / _` |/ _ \  | __/ _ \ / _ \| / __| | |/ _ \/ _` |
  | |    | (_| | (_) | | || (_) | (_) | \__ \ | |  __/ (_| |
  |_|     \__, |\___/   \__\___/ \___/|_|___/ |_|\___|\__,_|
          |___/
  https://uninformativ.de/git/r-go-tools-led
  https://uninformativ.de/bugs.html


- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -


A set of tools to work with the special LED on R-Go Tools Split
keyboards on GNU/Linux.


Installation
------------

r-go-led, the program to set the state of the LED, uses the hidraw
interface of Linux. This means you need to have the Linux API headers
for your kernel installed during compilation.

To compile it, run:

    $ make

The script r-go-break-reminder needs xprintidle (there is a package for
it on Arch Linux and Ubuntu and possible other distributions as well):

    - https://github.com/g0hl1n/xprintidle

r-go-break-reminder expects that both xprintidle and r-go-led are
available in your $PATH.

r-go-led needs access to /dev/hidraw* during runtime. For quick tests,
using sudo is a good option. A better setup might be using a dedicated
system group and a udev rule such as:

    KERNEL=="hidraw*", \
        ATTRS{idVendor}=="0911", ATTRS{idProduct}=="2188", \
        GROUP="rgoled"

This makes the device nodes accessible to the group "rgoled". You can
now install the r-go-led binary as group "rgoled" and with the SGID bit
set to grant it the appropriate permissions during runtime.


Running
-------

r-go-led is meant to be run from the command line or from other scripts.
Refer to the manpage for a full list of options. For example, this sets
the LED to green:

    $ r-go-led -g

r-go-break-reminder detects long streaks of user activity and then
activates the LED, until the user took a break for a while. You should
probably launch it from your ~/.xinitrc. Again, the manpage lists all
options.


CHANGES
Release history for r-go-tools-led

v21.02  2021-02-24
  [Changed]
  - r-go-break-reminder: "-s" now needs an argument: The command used to
    elevate privileges. This is usually "sudo" or "doas".

v21.01  2021-01-17
  [Changed]
  - r-go-break-reminder: LED will be set to green, if configured idle
    time has been reached. It goes back to "off" during the "working"
    period. And it's red, as before, if you should take a break.

v20.12  2020-12-05
  - Initial release.
  - r-go-led, r-go-break-reminder.


r-go-led.c
#define _POSIX_C_SOURCE 2  /* for getopt() */

#include <fcntl.h>
#include <glob.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

#include <linux/types.h>
#include <linux/input.h>
#include <linux/hidraw.h>

enum RGoLEDColor
{
    RGO_LED_COLOR_RED,
    RGO_LED_COLOR_GREEN,
    RGO_LED_COLOR_YELLOW,
    RGO_LED_COLOR_OFF,
};

void
set_leds_on_hidraw(char *path, int16_t vendor, int16_t product,
                   enum RGoLEDColor rgo_color)
{
    int fd;
    struct hidraw_devinfo info = {0};
    uint8_t buf[8] = {0};
    uint8_t color_byte = 0;

    fd = open(path, O_RDONLY);
    if (fd == -1)
    {
        fprintf(stderr, "open %s: ", path);
        perror("");
        return;
    }

    if (ioctl(fd, HIDIOCGRAWINFO, &info) < 0)
    {
        fprintf(stderr, "ioctl for dev info %s: ", path);
        perror("");
        close(fd);
        return;
    }

    if (info.vendor == vendor && info.product == product)
    {
        switch (rgo_color)
        {
            case RGO_LED_COLOR_RED:    color_byte = 0x01; break;
            case RGO_LED_COLOR_GREEN:  color_byte = 0x02; break;
            case RGO_LED_COLOR_YELLOW: color_byte = 0x03; break;
            case RGO_LED_COLOR_OFF:    color_byte = 0x04; break;
        }

        /* This ioctl() sends a "feature report" to the device.
         *
         * hidraw devices allow us to send control transfer requests to
         * endpoint 0 *without* having to detach the kernel driver. We
         * can't send arbitrary data, but it's good enough to control
         * the special R-Go Tools LED on the keyboard.
         *
         * bmRequestType: Fixed at 0x21 by Linux
         * bRequest: Fixed at 0x09 by Linux
         * wValue: High byte fixed at 0x03 by Linux (determined by
         *         the fact that we call the ioctl() to do a feature
         *         report)
         * wValue: Low byte (report ID) is the first byte from the
         *         payload (0x30) and it's also duplicated into the
         *         the actual payload
         * wIndex: Fixed by Linux, depends on which device we use
         *         ("interface_number")
         *
         * Remaining payload (0x91, ...) as documented by manufacturer.
         * The trailing 0x00 don't appear to be strictly necessary, but
         * that's what they told me to use, so we'll to that.
         */
        buf[0] = 0x30;
        buf[1] = 0x91;
        buf[2] = color_byte;
        buf[3] = 0x00;
        buf[4] = 0x00;
        buf[5] = 0x00;
        buf[6] = 0x00;
        buf[7] = 0x00;
        if (ioctl(fd, HIDIOCSFEATURE(8), buf) < 0)
        {
            fprintf(stderr, "ioctl for sending feature report to %s: ", path);
            perror("");
        }
    }

    close(fd);
}

void
set_leds(int16_t vendor, int16_t product, enum RGoLEDColor rgo_color)
{
    glob_t globres;
    size_t path_i;

    glob("/dev/hidraw*", GLOB_NOSORT, NULL, &globres);
    for (path_i = 0; path_i < globres.gl_pathc; path_i++)
    {
        set_leds_on_hidraw(globres.gl_pathv[path_i], vendor, product, rgo_color);
    }
    globfree(&globres);
}

int
main(int argc, char **argv)
{
    enum RGoLEDColor rgo_color = RGO_LED_COLOR_OFF;
    int16_t vendor = 0x0911, product = 0x2188;
    int opt;

    while ((opt = getopt(argc, argv, "rgyv:p:")) != -1)
    {
        switch (opt)
        {
            case 'r': rgo_color = RGO_LED_COLOR_RED; break;
            case 'g': rgo_color = RGO_LED_COLOR_GREEN; break;
            case 'y': rgo_color = RGO_LED_COLOR_YELLOW; break;
            case 'v': vendor = atoi(optarg); break;
            case 'p': product = atoi(optarg); break;
            default: return 1;
        }
    }

    if (vendor == 0 || product == 0)
    {
        fprintf(stderr, "Invalid vendor or product ID\n");
        return 1;
    }

    set_leds(vendor, product, rgo_color);

    return 0;
}


standard-leds.c
#define _POSIX_C_SOURCE 2  /* for getopt() */

#include <fcntl.h>
#include <glob.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

#include <linux/types.h>
#include <linux/input.h>
#include <linux/hidraw.h>

uint8_t STANDARD_LED_NUMLOCK = 1;
uint8_t STANDARD_LED_CAPSLOCK = 2;
uint8_t STANDARD_LED_SCROLLLOCK = 4;

void
set_leds_on_hidraw(char *path, int16_t vendor, int16_t product,
                   uint8_t standard_led_mask)
{
    int fd;
    struct hidraw_devinfo info = {0};
    uint8_t buf[2] = {0};

    fd = open(path, O_RDWR);
    if (fd == -1)
    {
        fprintf(stderr, "open %s: ", path);
        perror("");
        return;
    }

    if (ioctl(fd, HIDIOCGRAWINFO, &info) < 0)
    {
        fprintf(stderr, "ioctl for dev info %s: ", path);
        perror("");
        close(fd);
        return;
    }

    if (info.vendor == vendor && info.product == product)
    {
        /* To change the standard LEDs, we'll send an "output
         * report". This is a control transfer.
         *
         * hidraw lets use do this, too, but be aware that the kernel
         * driver keeps an internal state of the LEDs. It won't know
         * that we changed it and will overwrite us the next time the
         * user hits a key like numlock. Also, changing the LED does not
         * affect the actual state of that modifier, e.g. changing the
         * LED of caps lock does not activate caps lock on your Linux
         * host -- but it MIGHT trigger something in the keyboard.
         * Nobody knows.
         *
         * tl;dr: Use with caution.
         *
         * bmRequestType: Fixed at 0x21 by Linux
         * bRequest: Fixed at 0x09 by Linux
         * wValue: High byte fixed at 0x02 by Linux (determined by
         *         the fact that we call "write()")
         * wValue: Low byte (report ID) is the first byte from the
         *         payload (0x00), NOT duplicated into real payload
         * wIndex: Fixed by Linux, depends on which device we use
         *         ("interface_number")
         *
         * Remaining payload indicates state of each LED:
         *
         * - Bit 0: Numlock
         * - Bit 1: Caps lock
         * - Bit 2: Scroll lock
         *
         * Changing these standard LEDs is documented in numerous
         * places:
         *
         * https://www.usb.org/sites/default/files/documents/hid1_11.pdf
         * https://wiki.osdev.org/USB_Human_Interface_Devices#LED_lamps
         */
        buf[0] = 0x00;
        buf[1] = standard_led_mask;
        if (write(fd, buf, 2) < 0)
        {
            fprintf(stderr, "ioctl for sending output report to %s: ", path);
            perror("");
        }
    }

    close(fd);
}

void
set_leds(int16_t vendor, int16_t product, uint8_t standard_led_mask)
{
    glob_t globres;
    size_t path_i;

    glob("/dev/hidraw*", GLOB_NOSORT, NULL, &globres);
    for (path_i = 0; path_i < globres.gl_pathc; path_i++)
    {
        set_leds_on_hidraw(globres.gl_pathv[path_i], vendor, product,
                           standard_led_mask);
    }
    globfree(&globres);
}

int
main(int argc, char **argv)
{
    uint8_t standard_led_mask = 0;
    int16_t vendor = 0x0911, product = 0x2188;
    int opt;

    while ((opt = getopt(argc, argv, "ncsv:p:")) != -1)
    {
        switch (opt)
        {
            case 'n': standard_led_mask |= STANDARD_LED_NUMLOCK; break;
            case 'c': standard_led_mask |= STANDARD_LED_CAPSLOCK; break;
            case 's': standard_led_mask |= STANDARD_LED_SCROLLLOCK; break;
            case 'v': vendor = atoi(optarg); break;
            case 'p': product = atoi(optarg); break;
            default: return 1;
        }
    }

    if (vendor == 0 || product == 0)
    {
        fprintf(stderr, "Invalid vendor or product ID\n");
        return 1;
    }

    set_leds(vendor, product, standard_led_mask);

    return 0;
}
